core/modules/link/link.module

```diff
@@ -849,7 +849,27 @@ function _link_default_attributes() {
  * The $fields array is in $complete_form['#field_info'][$element['#field_name']].
  */
 function link_field_process($element, $form_state, $complete_form) {
-  $instance = field_widget_instance($element, $form_state);
+  $field_parents = (!empty($element['#field_parents'])) ? $element['#field_parents'] : array();
+  $field_name = (!empty($element['#field_name'])) ? $element['#field_name'] : '';
+  $language = (!empty($element['#language'])) ? $element['#language'] : LANGUAGE_NONE;
+  $field_state = field_form_get_state($field_parents, $field_name, $language, $form_state);
+  $instance = !empty($field_state['instance']) ? $field_state['instance'] : array();
+  $delta = !empty($element['#delta']) ? $element['#delta'] : 0;
+
+  if (!$instance) {
+    // The element comes from a custom form, we'll check for custom instance
+    // settings or use defaults.
+    $link_field_info = link_field_info();
+    $default_settings = $link_field_info['link_field']['instance_settings'];
+    $instance = (!empty($element['#instance'])) ? $element['#instance'] : array();
```

Comment:
I think this makes things a bit more cumbersome than your first commit. We don't use `#instance` as a FormAPI key for any other custom fields. It seems much neater to map element keys individually to me.

Reply:
I've made some changes in the PR.